src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs

```diff
@@ -656,6 +656,77 @@ private static bool VerifyGetAwaiter(IMethodSymbol getAwaiter)
             return methods.Any(m => m.Name == WellKnownMemberNames.GetResult && !m.Parameters.Any());
         }
 
+        public static bool IsValidGetEnumerator(this IMethodSymbol symbol)
+            => symbol.Name == WellKnownMemberNames.GetEnumeratorMethodName &&
+            VerifyGetEnumerator(symbol);
+
+        private static bool VerifyGetEnumerator(IMethodSymbol getEnumerator)
+        {
+            var returnType = getEnumerator.ReturnType;
+            if (returnType == null)
+            {
+                return false;
+            }
+
+            var members = returnType.AllInterfaces.Concat(returnType.GetBaseTypesAndThis())
+                .SelectMany(x => x.GetMembers())
+                .Where(x => x.DeclaredAccessibility == Accessibility.Public)
+                .ToList();
+
+            // T Current { get }
+            if (!members.OfType<IPropertySymbol>().Any(p => p.Name == WellKnownMemberNames.CurrentPropertyName && p.GetMethod != null))
+            {
+                return false;
+            }
+
+            // bool MoveNext()
+            if (!members.OfType<IMethodSymbol>().Any(x => x is
+            {
+                Name: WellKnownMemberNames.MoveNextMethodName,
+                ReturnType: { SpecialType: SpecialType.System_Boolean },
+                Parameters: { Length: 0 },
+            }))
```

Comment:
Indent this. Or extract local

Reply:
I get a formatting error when I indent it. Extracting a local makes it ugly enough I just wouldn't use patterns at all.